rbd: refactor mirroring to work with volume and volumegroup

Madhu-1 opened the pull request:
refactor mirroring to work with volume and volumegroup

Madhu-1 commented:
@nixpanic can you please check this PR for early review, i still need to do testing, once everything is good i will move this to Ready state.

nixpanic commented on scripts/golangci.yml.in:
```diff
@@ -208,3 +208,5 @@ linters:
     - containedctx
     # TODO: depguard requires a list of (un)acceptable imports
     - depguard
+    # Disable interfacebloat check as it expects onlt 10 methods in a interface
```
please disable this only with a comment on the interface that is affected

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -124,23 +126,23 @@ func getForceOption(ctx context.Context, parameters map[string]string) (bool, er
 // getFlattenMode gets flatten mode from the input GRPC request parameters.
 // flattenMode is the key to check the mode in the parameters.
-func getFlattenMode(ctx context.Context, parameters map[string]string) (corerbd.FlattenMode, error) {
+func getFlattenMode(ctx context.Context, parameters map[string]string) (string, error) {
```
Why remove the FlattenMode type? You can put it in a file under `internal/rbd/types` if that makes more sense.

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -422,37 +435,44 @@ func (rs *ReplicationServer) PromoteVolume(ctx context.Context,
+	var rbdVol types.Volume
 
-		return nil, err
+	if req.GetReplicationSource().GetVolume() != nil {
```
can `req.GetReplicationSource()` be nil?

nixpanic commented on internal/rbd/mirror.go:
```diff
@@ -42,13 +44,17 @@ const (
 func (rv *rbdVolume) HandleParentImageExistence(
 	ctx context.Context,
-	flattenMode FlattenMode,
+	flattenMode string,
```
don't move to a generic string, it is much better to keep the FlattenMode type.

nixpanic commented on internal/rbd/mirror.go:
```diff
@@ -88,8 +99,11 @@ func (rv *rbdVolume) HandleParentImageExistence(
-// EnableImageMirroring enables mirroring on an image.
-func (ri *rbdImage) EnableImageMirroring(mode librbd.ImageMirrorMode) error {
+// check that rbdVolume implements the types.Mirror interface.
+var _ types.Volume = &rbdVolume{}
```
use `types.Mirror` ?

nixpanic commented on internal/rbd/types/mirror.go:
```diff
@@ -0,0 +1,52 @@
+	// GetLocalState returns the local state of the resource
+	GetLocalState() (interface{}, error)
+
+	HandleParentImageExistence(ctx context.Context, flattenMode string) error
```
missing comments for these functions

nixpanic commented on internal/rbd/types/volume.go:
```diff
@@ -34,4 +35,11 @@ type Volume interface {
 	// ToCSI creates a CSI protocol formatted struct of the volume.
 	ToCSI(ctx context.Context) *csi.Volume
+
+	GetPoolName() string
```
please add comments for these functions too

nixpanic commented on internal/rbd/types/volume.go:
```diff
@@ -34,4 +35,11 @@ type Volume interface {
+	GetCreationTime() (*timestamppb.Timestamp, error)
+	GetMetadata(key string) (string, error)
+	SetMetadata(key, value string) error
+	RepairResyncedImageID(ctx context.Context, ready bool) error
```
`RepairResyncedImageID` looks more like a function that belongs in the `Mirror` type?

Madhu-1 replied:
This is to update the imageID in volume OMAP, I will recheck in the next followup PR and move this to Mirror based on the requirement.

nixpanic commented on internal/rbd/types/mirror.go:
```diff
@@ -0,0 +1,52 @@
+	"github.com/ceph/go-ceph/rbd/admin"
+)
+
+type Mirror interface {
```
Consider calling this `MirroredVolume` maybe?

Ideally each new interface type has a `Destroy(context.Context)` function, and all other functions get a context.Context parameter as well. That makes all interfaces look more similar, and easier to use/maintain.

Madhu-1 replied:
I called this as Mirror as its going to be common for both volume and group, let me know if this is fine or any other suggestion, I will keep this in Mind and add Destory and get in next implementation for now it looks like its not required and they are already part of volume and group which you are working on.

nixpanic replied:
Ok, `Mirror` is fine. We can always change it later it makes more sense then.

Yes, `Destroy()` is already part of the other types, so only required in case a function allocates/returns a `Mirror` type.

nixpanic commented on internal/rbd/types/mirror.go:
```diff
@@ -0,0 +1,52 @@
+	Resync() error
+
+	// GetMirroringInfo returns the mirroring information of the resource
+	GetMirroringInfo() (interface{}, error)
```
It is really ugly to return an untyped interface. Can that be improved?

Madhu-1 replied:
i tried but it gets more difficult for implementation where i need to have different duplicate checks for both volume and group, that's the  reason to settle on this after lot of other considerations

nixpanic replied:
Maybe add a `MirrorInfo` interface too?

Madhu-1 replied:
Am not sure adding MirrorInfo interface helps but the problem is that the volumemirror and groupmirror in go-ceph return different structs, if both are returning the same i could have solved this or else I need to own a local struct and fill the required values from both, i can do this if you prefer (this will have lot of duplication logic and maintenance work)

nixpanic replied:
Maybe, or have a MirrorInfo interface with only some basic function, or completely empty? At least the type can then describe that it is actually one of the go-ceph types.

nixpanic commented on internal/rbd/types/volume.go:
```diff
@@ -20,8 +20,10 @@ import (
+	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
+//nolint:interfacebloat // more than 10 methods are needed for the interface
```
In #4719  I added the `journalledObject` interface. Such an approach also prevents the need for skipping the `interfacebloat` linter.

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -270,24 +272,26 @@ func (rs *ReplicationServer) EnableVolumeReplication(ctx context.Context,
-			rbdVol.Destroy(ctx)
-		}
-	}()
+	mgr := rbd.NewManager(req.GetParameters(), req.GetSecrets())
```
note that the CSI-instance id (drivername) needs to be passed to the manager in an upcoming change.

Madhu-1 commented:
@nixpanic I have tested this code and works as expected, let me know if you want to make any changes you are referring to the manager as part of this PR.

nixpanic commented on internal/rbd/mirror.go:
```diff
@@ -223,26 +219,104 @@ func (ri *rbdImage) GetImageMirroringStatus() (*librbd.GlobalMirrorImageStatus,
-func (ri *rbdImage) GetLocalState() (librbd.SiteMirrorImageStatus, error) {
-	localStatus := librbd.SiteMirrorImageStatus{}
-	image, err := ri.open()
+type ImageStatus struct {
```
A description of what this type represents would be good.

nixpanic commented on internal/rbd/mirror.go:
```diff
@@ -223,26 +219,104 @@ func (ri *rbdImage) GetImageMirroringStatus() (*librbd.GlobalMirrorImageStatus,
+	return SiteMirrorImageStatus{SiteMirrorImageStatus: ss}, err
+}
+
+type SiteMirrorImageStatus struct {
```
A description of what this type represents would be good.

nixpanic commented on internal/rbd/types/mirror.go:
```diff
@@ -0,0 +1,87 @@
+	FlattenModeForce FlattenMode = "force"
+)
+
+type Mirror interface {
```
A description of what this type represents would be good. I think this can be either a single RBD-image or an RBD-group?

nixpanic commented on internal/rbd/types/mirror.go:
```diff
@@ -0,0 +1,87 @@
+	AddSnapshotScheduling(interval admin.Interval, startTime admin.StartTime) error
+}
+
+type MirrorInfo interface {
```
A description of what this type represents would be good. Is this the type that describes the local status, or the remote status?

Madhu-1 replied:
This is for the image local state but this is not part of local status.

nixpanic commented on internal/rbd/types/mirror.go:
```diff
@@ -0,0 +1,87 @@
+	// IsPrimary returns true if the resource is primary
+	IsPrimary() bool
+	// GetState returns the state of the resource
+	GetState() string
```
consider adding a `MirrorState` type to replace the string value

Madhu-1 replied:
This needs more work, will take it in next update when implementing this for group

nixpanic commented on internal/rbd/types/volume.go:
```diff
@@ -40,4 +42,23 @@ type Volume interface {
+	// if the parent image exists and is not enabled for mirroring, it returns an error.
+	HandleParentImageExistence(ctx context.Context, flattenMode FlattenMode) error
+
+	Mirror
```
Comment on this please, and move it to the top so that it is immediately clear that everything from `Mirror` is inherited too.

Or, maybe introduce a `MirroredVolume` type that inherits both from `Volume` and `Mirror`? That way the `Volume` interface stays small and simple for 'normal' usage, the `rbdVolume` object can still implement both.

Madhu-1 replied:
Done. Moved it to top.

nixpanic reviewed (commented):
Splitting the commit would be nice.
At least add the new types in a separate commit, and use them in subsequent ones.

The single commit is +488 -300 lines of code, which is rather a lot.

Madhu-1 committed:
rbd: add mirror interface

Adding mirror interface for rbd image
and group which can be used to manage
mirroring for both.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: add mirroring helper for image

Add helper methods for the rbd
image mirroring.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: refractor to use mirror interface

Refractoring code to use mirror
interface.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented on internal/rbd/types/mirror.go:
```diff
@@ -0,0 +1,97 @@
+
+// MirrorImage is the interface for managing mirroring on an RBD image or group of images.
+// This will be used to get the state of resource and it is primary or secondary.
+
```
nit: unneeded empty line

nixpanic commented on internal/rbd/types/mirror.go:
```diff
@@ -0,0 +1,97 @@
+
+// GlobalStatus is the interface for fetching the global status of the mirroring.
+// This will be used to get the status of the local site and remote site or all sites.
+
```
nit: unneeded empty line

nixpanic commented on internal/rbd/types/volume.go:
```diff
@@ -40,4 +42,21 @@ type Volume interface {
 	RemoveFromGroup(ctx context.Context, vg VolumeGroup) error
+
+	// GetPoolName returns the name of the pool where the volume is stored.
+	GetPoolName() string
```
Maybe `Volume` should inherit the `journalledObject` too?

https://github.com/ceph/ceph-csi/blob/aa88b4c4a01554db6456b971fe387dbd391d0a49/internal/rbd/types/group.go#L26-L38

Madhu-1 replied:
This needs a little more refactoring as journalledObject is specific to volumeGroup now, will address this in follow-up when implementing for group support

nixpanic replied:
It is only the comments that refer to VolumeGroup. The functions could be for Volume as well.

nixpanic commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -270,24 +277,25 @@ func (rs *ReplicationServer) EnableVolumeReplication(ctx context.Context,
 
 		return nil, err
 	}
+	mirror = rbdVol
```
Would it make sense to have `rbdVol.ToMirror()` that validates everything required for the new Mirror type?

Madhu-1 replied:
@nixpanic can you explain a bit more on this one?

Madhu-1 replied:
updated the code, PTAL and see if this matches your expectation

nixpanic commented on internal/rbd/types/volume.go:
```diff
@@ -25,6 +25,8 @@ import (
 //nolint:interfacebloat // more than 10 methods are needed for the interface
 type Volume interface {
+	// Mirror interface implementation for Volume
+	Mirror
```
because you use `mirror = rbdVol`, I would expect that the `Mirror` interface is kept seperate. I do not see a need to have it part of the Volume interface. The `rbdVolume` struct can still implement both interfaces.

nixpanic commented on internal/rbd/group.go:
```diff
@@ -77,3 +77,7 @@ func (rv *rbdVolume) RemoveFromGroup(ctx context.Context, vg types.VolumeGroup)
 	return librbd.GroupImageRemove(ioctx, name, rv.ioctx, rv.RbdImageName)
 }
+
+func (rv *rbdVolume) ToMirror() types.Mirror {
```
Thanks! Would it make sense to have some checking if mirror attributes are valid?

Madhu-1 replied:
Not for now, do you have any specific checks in your mind?

nixpanic replied:
Not really. Mostly just to have the function match `func (*rbdVolume) ToCSI(ctx) (csi.Volume, error)`

Madhu-1 replied:
the signature is matching now, it returns nil for now but can be updated later as required.

nixpanic commented:
@mergifyio rebase

iPraveenParihar commented on internal/rbd/types/mirror.go:
```diff
@@ -0,0 +1,95 @@
+	AddSnapshotScheduling(interval admin.Interval, startTime admin.StartTime) error
+}
+
+// MirrorImage is the interface for managing mirroring on an RBD image or group of images.
```
```suggestion
// MirrorInfo is the interface for managing mirroring on an RBD image or group of images.
```

Madhu-1 replied:
Will address this in follow up. Thanks @Madhu-1 for catching it.

iPraveenParihar reviewed (approved):
nit

iPraveenParihar commented on internal/rbd/driver/driver.go:
```diff
@@ -219,7 +219,7 @@ func (r *Driver) setupCSIAddonsServer(conf *util.Config) error {
 		r.cas.RegisterService(fcs)
 
-		rcs := casrbd.NewReplicationServer(NewControllerServer(r.cd))
+		rcs := casrbd.NewReplicationServer(rbd.CSIInstanceID, NewControllerServer(r.cd))
```
@Madhu-1 , shouldn't this be
```suggestion
rcs := casrbd.NewReplicationServer(conf.InstanceID, NewControllerServer(r.cd))
```
?

nixpanic replied:
yes, I think so. Can you send a PR for that?

Madhu-1 replied:
`conf.InstanceID` will not work if its not set, we need to use default in that case we are not updating conf.InstanceID with rbd.CSIInstanceID, that's the reason to use `rbd.CSIInstanceID` as its gets updated always or it contains the default

iPraveenParihar replied:
@Madhu-1, we are doing that for InitJournals - 
https://github.com/ceph/ceph-csi/blob/db6f54f75d87261d66cada71e6a2a0f8665cec18/internal/rbd/globals.go#L93-L100

And `conf.InstanceID` is passed to InitJournals

Madhu-1 replied:
if the conf.InsatceID is empty its not updated with the default one that we have in the code right rbd.InitJournals(conf.InstanceID)? with this one https://github.com/ceph/ceph-csi/blob/db6f54f75d87261d66cada71e6a2a0f8665cec18/internal/rbd/globals.go#L28

iPraveenParihar replied:
If `conf.InstanceID` is empty,  `CSIInstanceID` (default) is used for journals. Check the above code I have linked in previous comment.

iPraveenParihar replied:
Apologies, I was mistaken it was other way around. 
Thanks @Madhu-1.

Then I guess for VolumeGroupServer we need to use `rbd.CSIInstanceID` 
https://github.com/ceph/ceph-csi/blob/db6f54f75d87261d66cada71e6a2a0f8665cec18/internal/rbd/driver/driver.go#L223
@nixpanic ?

nixpanic replied:
`rbd.CSIInstanceID` seems to be a global value, it is configurable on the commandline, so I think using the configured value is cleaner (all config options should have default values anyway?).

In `internal/rbd/driver/driver.go` there is `rbd.InitJournals(conf.InstanceID)`, so in the end it is all the same.

iPraveenParihar replied:
`conf.InstanceID` default value is empty and `rbd.CSIInstanceID` is replaced with `conf.InstanceID` when it is configured in command line.
So using `conf.InstanceID` when its not configured  will lead to creating journal directory `csi.groups.` ?

nixpanic replied:
yeah, I noticed that too. The global `rbd.CSIInstanceID` could be set to something earlier on, in case the ordering of (seemingly unrelated) initialization of the journals is done later.  This is a little fragile, and using global variables is quite ugly in any case. I'm preparing a PR for improving it.

nixpanic replied:
see #4747

Rakshith-R commented on internal/csi-addons/rbd/replication.go:
```diff
@@ -270,24 +277,27 @@ func (rs *ReplicationServer) EnableVolumeReplication(ctx context.Context,
-	rbdVol, err := corerbd.GenVolFromVolID(ctx, volumeID, cr, req.GetSecrets())
-	defer func() {
-		if rbdVol != nil {
-			rbdVol.Destroy(ctx)
```
I guess this got missed during refractor?

Madhu-1 replied:
yes this is a miss
